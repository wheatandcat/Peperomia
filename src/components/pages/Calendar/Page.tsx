import React, { memo, useCallback } from 'react';
import {
  View,
  FlatList,
  ListRenderItemInfo,
  StyleSheet,
  TouchableOpacity,
} from 'react-native';
import dayjs from 'dayjs';
import advancedFormat from 'dayjs/plugin/advancedFormat';
import 'dayjs/locale/ja';
import CreateHeader from 'components/organisms/Header/CreateHeader.tsx';
import Header from 'components/organisms/Header/Header.tsx';
import { getKindData } from 'lib/kind';
import Card from 'components/organisms/Calendar/Card.tsx';
import Add from 'components/organisms/Calendar/Add.tsx';
import theme from 'config/theme';
import ItemWrap from 'components/organisms/ItemWrap/ItemWrap';
import { ConnectedType, CalendarType, ItemDetailType } from './Connected';

dayjs.extend(advancedFormat);

type Props = ConnectedType & {
  calendar: CalendarType;
};

type ItemDetailType2 = ItemDetailType & {
  add?: boolean;
};

type Item = {
  item1: ItemDetailType2;
  item2: ItemDetailType2;
};

const CalendarPage: React.FC<Props> = (props) => {
  const calendar = props.calendar;
  const itemDetails = calendar.item.itemDetails || [];

  const config = getKindData(calendar.item.kind || '');

  const renderItem = useCallback(
    ({ item }: ListRenderItemInfo<Item>) => {
      return (
        <View style={styles.cards}>
          {(() => {
            if (item?.item1?.add) {
              return (
                <TouchableOpacity
                  onPress={props.onAddItemDetail}
                  style={styles.card1}
                >
                  <View>
                    <Add />
                  </View>
                </TouchableOpacity>
              );
            }

            if (item.item1) {
              return (
                <TouchableOpacity
                  onPress={() => props.onItemDetail(item.item1.id)}
                  style={styles.card1}
                >
                  <Card {...item.item1} />
                </TouchableOpacity>
              );
            }

            return null;
          })()}
          {(() => {
            if (item?.item2?.add) {
              return (
                <TouchableOpacity
                  onPress={props.onAddItemDetail}
                  style={styles.card2}
                >
                  <View>
                    <Add />
                  </View>
                </TouchableOpacity>
              );
            }

            if (item.item2) {
              return (
                <TouchableOpacity
                  onPress={() => props.onItemDetail(item.item2.id)}
                  style={styles.card2}
                >
                  <Card {...item.item2} />
                </TouchableOpacity>
              );
            }

            return null;
          })()}
        </View>
      );
    },
    [props]
  );

  const listItem = [
    ...itemDetails.slice(1),
    {
      add: true,
      id: 'add-item',
      title: '',
      kind: '',
    },
  ];

  const list1 = listItem.filter((_, index) => {
    return index % 2 === 0;
  });
  const list2 = listItem.filter((_, index) => {
    return index % 2 === 1;
  });

  const data = list1.map((_, index) => ({
    item1: list1[index] || null,
    item2: list2[index] || null,
  })) as Item[];

  return (
    <ItemWrap kind={calendar.item.kind || ''}>
      <FlatList<Item>
        keyExtractor={(_, index) => `card_${index}`}
        data={data}
        renderItem={renderItem}
        ListHeaderComponent={
          <View style={styles.header}>
            {props.create ? (
              <CreateHeader
                color={config.backgroundColor}
                date={dayjs(calendar.date).format('YYYY年MM月DD日')}
                title={calendar.item.title || ''}
                kind={calendar.item.kind || ''}
                onClose={props.onDismiss}
              />
            ) : (
              <Header
                color={config.backgroundColor}
                date={dayjs(calendar.date).format('YYYY年MM月DD日')}
                title={calendar.item.title || ''}
                kind={calendar.item.kind || ''}
                public={calendar.public}
                onClose={props.onDismiss}
                onUpdate={props.onUpdate}
                onShare={props.onShare}
                onDelete={props.onDelete}
              />
            )}
          </View>
        }
        ListFooterComponent={<View style={styles.footer} />}
        showsVerticalScrollIndicator={false}
        showsHorizontalScrollIndicator={false}
        contentContainerStyle={styles.contentContainerStyle}
      />
    </ItemWrap>
  );
};

export default memo(CalendarPage);

const styles = StyleSheet.create({
  header: {
    paddingBottom: theme().space(2),
  },
  footer: {
    width: '100%',
    height: '100%',
    backgroundColor: theme().color.white,
  },
  cards: {
    width: '100%',
    flexDirection: 'row',
    backgroundColor: theme().color.white,
  },
  card1: {
    width: '50%',
    padding: theme().space(2),
    paddingRight: theme().space(2),
  },
  card2: {
    width: '50%',
    padding: theme().space(2),
    paddingRight: theme().space(2),
  },
  contentContainerStyle: {
    flex: 1,
  },
});
