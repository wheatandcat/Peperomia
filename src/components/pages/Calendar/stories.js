import React from 'react';
import { View, StyleSheet } from 'react-native';
import { storiesOf } from '@storybook/react-native';
import dayjs from 'dayjs';
import Page from './Page';

storiesOf('pages', module).add('Calendar', () => (
  <View style={styles.root}>
    <Page
      title="青山ランチ"
      date={dayjs().format('YYYY年MM月DD日')}
      kind="lunch"
    />
  </View>
));

const styles = StyleSheet.create({
  root: {
    paddingTop: 30,
  },
});
