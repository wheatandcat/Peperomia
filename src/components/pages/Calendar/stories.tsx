import React from 'react';
import { View, StyleSheet } from 'react-native';
import { storiesOf } from '@storybook/react-native';
import { KIND_PARK, KIND_AQUARIUM } from 'peperomia-util';
import { mockFn, StackNavigator } from 'storyBookUtils';
import Page from './Page';

const data = {
  calendar: {
    id: '1',
    date: '2020-01-01 00:00:00',
    public: true,
    item: {
      id: '1',
      title: '青山ランチ',
      kind: 'park',
      itemDetails: [
        {
          id: '1',
          title: '新宿駅',
          kind: KIND_PARK,
          memo: '',
          url: '',
          place: '',
          priority: 1,
        },
        {
          id: '2',
          title: '水族館',
          kind: KIND_AQUARIUM,
          memo: '',
          url: '',
          place: '',
          priority: 1,
        },
        {
          id: '3',
          title: '水族館',
          kind: KIND_AQUARIUM,
          memo: '',
          url: '',
          place: '',
          priority: 1,
        },
      ],
    },
  },
};

const Screen = () => {
  return (
    <View style={styles.root}>
      <Page
        calendar={data.calendar}
        onDismiss={mockFn('onDismiss')}
        onUpdate={mockFn('onUpdate')}
        onDelete={mockFn('onDelete')}
        onShare={mockFn('onShare')}
        onAddItemDetail={mockFn('onAddItemDetail')}
        onItemDetail={mockFn('onItemDetail')}
        create={false}
      />
    </View>
  );
};

storiesOf('pages', module).add('Calendar', () => (
  <StackNavigator screen={Screen} />
));

const styles = StyleSheet.create({
  root: {
    paddingTop: 0,
  },
});
