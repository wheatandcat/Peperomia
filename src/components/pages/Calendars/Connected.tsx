import React, { memo, useCallback } from 'react';
import dayjs from 'dayjs';
import advancedFormat from 'dayjs/plugin/advancedFormat';
import 'dayjs/locale/ja';
import { ContextProps as CalendarsContextProps } from 'containers/Calendars';
import { Props as IndexProps } from './';
import Page from './Page';

dayjs.extend(advancedFormat);

type Props = IndexProps &
  Pick<CalendarsContextProps, 'calendars' | 'setDate'> & {
    loading: boolean;
  };

export type ConnectedType = {
  onCreate: (date: string) => void;
  onCalendar: (date: string) => void;
  onSchedule: (id: string | number, title: string) => void;
};

const Connected: React.FC<Props> = memo((props) => {
  const onCreate = useCallback(
    (date: string) => {
      props.navigation.navigate('CreatePlan', {
        date,
      });
    },
    [props.navigation]
  );

  const onSchedule = useCallback(
    (id: string | number, title: string) => {
      props.navigation.navigate('Schedule', { itemId: id, title });
    },
    [props.navigation]
  );

  const onCalendar = useCallback(
    (date: string) => {
      props.navigation.navigate('Calendar', {
        date: dayjs(date).format('YYYY-MM-DDT00:00:00'),
      });
    },
    [props.navigation]
  );

  return (
    <Page
      calendars={props.calendars}
      setDate={props.setDate}
      loading={props.loading}
      onCreate={onCreate}
      onSchedule={onSchedule}
      onCalendar={onCalendar}
    />
  );
});

export default Connected;
