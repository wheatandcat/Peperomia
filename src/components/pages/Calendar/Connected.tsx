import React, { memo, useCallback } from 'react';
import { Alert, Dimensions } from 'react-native';
import { useDeleteCalendarMutation } from 'queries/api/index';
import { ContextProps as CalendarsContextProps } from 'containers/Calendars';
import { copyShareURL } from 'lib/share';
import {
  useUpdateCalendarPublicMutation,
  UpdateCalendarPublicMutationVariables,
} from 'queries/api/index';
import Toast from 'react-native-root-toast';
import useCalendar from 'hooks/useCalendar';
import { Props as IndexProps } from './';
import Plain, { QueryProps } from './Plain';

export type CalendarType = QueryData<QueryProps, 'calendar'>;
export type ItemDetailType = ArrayType<CalendarType['item']['itemDetails']>;

type Props = IndexProps & {
  date: string;
} & Pick<CalendarsContextProps, 'refetchCalendars'>;

export type ConnectedType = {
  onDismiss: () => void;
  onUpdate: () => void;
  onDelete: () => void;
  onShare: (open: boolean) => void;
  onAddItemDetail: () => void;
  onItemDetail: (itemDetailId: string) => void;
  create: boolean;
};

const Connected: React.FC<Props> = memo((props) => {
  const [updateCalendarPublicMutation] = useUpdateCalendarPublicMutation({
    async onCompleted(r) {
      const id = r.updateCalendarPublic.id;
      const open = r.updateCalendarPublic.public;

      if (open) {
        copyShareURL(String(id));
      } else {
        const { height } = Dimensions.get('window');

        let toast = Toast.show('リンクを非公開にしました', {
          duration: Toast.durations.LONG,
          position: height - 150,
          shadow: true,
          animation: true,
          hideOnPress: true,
          delay: 0,
        });

        setTimeout(function () {
          Toast.hide(toast);
        }, 3000);
      }
    },
    onError(err) {
      Alert.alert('公開に失敗しました', err.message);
    },
  });

  const [deleteCalendarMutation] = useDeleteCalendarMutation({
    async onCompleted() {
      await props.refetchCalendars?.();

      props.navigation.popToTop();
    },
    onError(err) {
      Alert.alert('削除に失敗しました', err.message);
    },
  });

  const { data, loading, error, refetch } = useCalendar({
    variables: {
      date: props.date,
    },
    fetchPolicy: 'network-only',
  });

  const onDismiss = useCallback(async () => {
    await props.refetchCalendars?.();

    props.navigation.popToTop();
  }, [props]);

  const onUpdate = useCallback(() => {
    const itemDetail = (data?.calendar?.item?.itemDetails || []).find(
      (v) => v?.priority === 1
    );

    props.navigation.navigate('EditItemDetail', {
      date: props.date,
      itemId: data?.calendar?.item.id || '',
      itemDetailId: itemDetail?.id || '',
      onCallback: async () => {
        await props.refetchCalendars?.();
        await refetch?.();
      },
    });
  }, [props, refetch, data]);

  const onDelete = useCallback(() => {
    const variables = {
      calendar: {
        date: props.date,
      },
    };

    deleteCalendarMutation({ variables });
  }, [deleteCalendarMutation, props.date]);

  const onAddItemDetail = useCallback(() => {
    props.navigation.navigate('AddItemDetail', {
      date: props.date,
      itemId: data?.calendar?.item.id || '',
      priority: (data?.calendar?.item.itemDetails?.length || 0) + 1,
      onCallback: async () => {
        await refetch?.();
      },
    });
  }, [props, data, refetch]);

  const onItemDetail = useCallback(
    (itemDetailId: string) => {
      props.navigation.navigate('ItemDetail', {
        date: props.date,
        itemId: data?.calendar?.item.id || '',
        itemDetailId,
        onCallback: async () => {
          await refetch?.();
        },
      });
    },
    [props, data, refetch]
  );

  const onShare = useCallback(
    async (open: boolean) => {
      const variables: UpdateCalendarPublicMutationVariables = {
        calendar: {
          date: props.date,
          public: open,
        },
      };
      updateCalendarPublicMutation({ variables });
    },
    [props, updateCalendarPublicMutation]
  );

  return (
    <Plain
      data={data}
      loading={loading}
      error={error}
      onDismiss={onDismiss}
      onAddItemDetail={onAddItemDetail}
      onUpdate={onUpdate}
      onDelete={onDelete}
      onItemDetail={onItemDetail}
      onShare={onShare}
      create={props.route.params.create || false}
    />
  );
});

export default Connected;
