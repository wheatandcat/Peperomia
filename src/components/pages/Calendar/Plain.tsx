import alertDialog from 'lib/alertDialog';
import ErrorPage from 'components/organisms/Error/Error';
import React from 'react';
import { CalendarQueryResult } from 'queries/api/index';
import Loading from 'components/atoms/Loading';
import { ConnectedType } from './Connected';
import Page from './Page';

export type QueryProps = Pick<
  CalendarQueryResult,
  'data' | 'loading' | 'error'
>;

type Props = QueryProps & ConnectedType;

const Plain: React.FC<Props> = (props) => {
  if (alertDialog(props)) return <ErrorPage {...props} />;
  if (props.loading) return <Loading loading />;

  const calendar = props.data?.calendar!;

  console.log(calendar);

  return <Page calendar={calendar} onDismiss={props.onDismiss} />;
};

export default Plain;
