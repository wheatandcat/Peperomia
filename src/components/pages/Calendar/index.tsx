import React, { memo } from 'react';
import { RouteProp } from '@react-navigation/native';
import { StackNavigationProp } from '@react-navigation/stack';
import { RootStackParamList } from 'lib/navigation';
import Connected from './Connected';

type ScreenNavigationProp = StackNavigationProp<RootStackParamList, 'Calendar'>;
type ScreenRouteProp = RouteProp<RootStackParamList, 'Calendar'>;

export type Props = {
  navigation: ScreenNavigationProp;
  route: ScreenRouteProp;
};

const Calendar: React.FC<Props> = memo((props) => {
  const date = props.route.params.date;

  return <Connected {...props} date={date} />;
});

export default Calendar;
