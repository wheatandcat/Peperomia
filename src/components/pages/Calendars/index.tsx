import React, { memo } from 'react';
import { RouteProp } from '@react-navigation/native';
import {
  createStackNavigator,
  StackNavigationProp,
} from '@react-navigation/stack';
import { RootStackParamList } from 'lib/navigation';
import { useCalendars } from 'containers/Calendars';
import Calendar from 'components/pages/Calendar/index';
import ItemDetail from 'components/pages/ItemDetail';
import AddItemDetail from 'components/pages/AddItemDetail/index';
import EditItemDetail from 'components/pages/EditItemDetail/index';
import { CreateCalendar } from 'components/pages/CreateCalendar/index';
import Icons, { IconsNavigationOptions } from 'components/pages/Icons';
import Connected from './Connected';

type CalendarsScreenNavigationProp = StackNavigationProp<
  RootStackParamList,
  'Calendars'
>;
type ScreenRouteProp = RouteProp<RootStackParamList, 'Calendars'>;

export type Props = {
  navigation: CalendarsScreenNavigationProp;
  route: ScreenRouteProp;
};

const Calendars: React.FC<Props> = memo((props) => {
  const { calendars, loadingCalendars, setDate } = useCalendars();

  return (
    <Connected
      {...props}
      calendars={calendars}
      loading={loadingCalendars}
      setDate={setDate}
    />
  );
});

const Stack = createStackNavigator<RootStackParamList>();

const RootStack = () => {
  return (
    <Stack.Navigator initialRouteName="Calendars" mode="modal">
      <Stack.Screen
        name="Calendars"
        component={Calendars}
        options={{ header: () => null }}
      />
      <Stack.Screen
        name="Calendar"
        component={Calendar}
        options={{ headerShown: false }}
      />
      <Stack.Screen
        name="CreateCalendar"
        component={CreateCalendar}
        options={{ header: () => null }}
      />
      <Stack.Screen
        name="ItemDetail"
        component={ItemDetail}
        options={{ header: () => null }}
      />
      <Stack.Screen
        name="AddItemDetail"
        component={AddItemDetail}
        options={{ header: () => null }}
      />
      <Stack.Screen
        name="EditItemDetail"
        component={EditItemDetail}
        options={{ header: () => null }}
      />
      <Stack.Screen
        name="Icons"
        component={Icons}
        options={IconsNavigationOptions}
      />
    </Stack.Navigator>
  );
};

export default RootStack;
