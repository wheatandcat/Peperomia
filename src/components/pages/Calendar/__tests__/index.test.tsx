import React from 'react';
import { shallow, ShallowWrapper } from 'enzyme';
import * as Calendars from 'containers/Calendars';
import * as Auth from 'containers/Auth';
import Calendar, { Props } from '../';

const propsData = (): Props =>
  ({
    navigation: {
      setParams: jest.fn(),
      navigate: jest.fn(),
    },
    route: {
      params: {
        date: '2020-01-01 00:00:00',
      },
    },
  } as any);

describe('components/pages/Calendar/index.tsx', () => {
  let wrapper: ShallowWrapper;

  beforeEach(() => {
    jest
      .spyOn(Calendars, 'useCalendars')
      .mockImplementation(() => ({ refetchCalendars: jest.fn() } as any));

    jest
      .spyOn(Auth, 'useAuth')
      .mockImplementation(() => ({ uid: 'test' } as any));

    wrapper = shallow(<Calendar {...propsData()} />);
  });

  it('正常にrenderすること', () => {
    expect(wrapper).toMatchSnapshot();
  });
});
