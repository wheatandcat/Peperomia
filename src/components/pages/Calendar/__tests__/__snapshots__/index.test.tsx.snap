// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`components/pages/Calendar 正常にrenderすること 1`] = `
<Memo()
  date="2020-01-01 00:00:00"
  navigation={
    Object {
      "navigate": [MockFunction],
      "setParams": [MockFunction],
    }
  }
  refetchCalendars={[MockFunction]}
  route={
    Object {
      "params": Object {
        "date": "2020-01-01 00:00:00",
      },
    }
  }
  uid="test"
/>
`;
